File: modules/GamePathManager.js
const path = require('path');
const fs = require('fs');
const fg = require('fast-glob');
const drivelist = require('drivelist');
const { exec } = require('child_process');

/**
 * 遊戲路徑檢測模塊
 * 負責自動檢測和管理 Blue Archive 遊戲路徑
 */
class GamePathManager {
    constructor(store, i18next) {
        this.store = store;
        this.i18next = i18next;
    }

    /**
     * 從 Steam 註冊檔獲取 Steam 安裝路徑
     */
    async getSteamInstallPath() {
        return new Promise((resolve) => {
            const registryPath = 'HKLM\\SOFTWARE\\WOW6432Node\\Valve\\Steam';
            const command = `reg query "${registryPath}" /v InstallPath`;
            exec(command, (error, stdout, stderr) => {
                if (error) {
                    console.error(`查詢 Steam 64位元登錄檔失敗: ${error.message}`);
                    const registryPath32 = 'HKLM\\SOFTWARE\\Valve\\Steam';
                    const command32 = `reg query "${registryPath32}" /v InstallPath`;
                    exec(command32, (error32, stdout32, stderr32) => {
                        if (error32) {
                            console.error(`查詢 Steam 32位元登錄檔失敗: ${error32.message}`);
                            resolve(null);
                            return;
                        }
                        const match32 = stdout32.match(/InstallPath\s+REG_SZ\s+(.*)/);
                        resolve(match32 ? match32[1].trim() : null);
                    });
                    return;
                }
                const match = stdout.match(/InstallPath\s+REG_SZ\s+(.*)/);
                resolve(match ? match[1].trim() : null);
            });
        });
    }

    /**
     * 通過 Steam 查找遊戲
     */
    async findGameViaSteam(win) {
        win.webContents.send('update-status', this.i18next.t('status_finding_steam'));
        const steamPath = await this.getSteamInstallPath();
        if (!steamPath) {
            console.log('在登錄檔中找不到 Steam 安裝路徑。');
            return null;
        }

        win.webContents.send('update-status', this.i18next.t('status_steam_found', { path: steamPath }));
        await new Promise(resolve => setTimeout(resolve, 1000));

        const libraryFoldersVdfPath = path.join(steamPath, 'steamapps', 'libraryfolders.vdf');
        if (!fs.existsSync(libraryFoldersVdfPath)) {
            console.log('找不到 libraryfolders.vdf 檔案。');
            return null;
        }

        try {
            const libraryFoldersContent = fs.readFileSync(libraryFoldersVdfPath, 'utf-8');
            const libraryPaths = [steamPath];
            const pathRegex = /"path"\s+"([^"]+)"/g;
            let match;
            while ((match = pathRegex.exec(libraryFoldersContent)) !== null) {
                const libPath = match[1].replace(/\\\\/g, '\\'); // 將 VDF 中的 \\ 轉為單一 \
                if (fs.existsSync(libPath)) {
                    libraryPaths.push(libPath);
                }
            }

            const uniqueLibraryPaths = [...new Set(libraryPaths)];
            console.log('找到的 Steam 遊戲庫:', uniqueLibraryPaths);

            const appId = '3557620'; // 蔚藍檔案的 Steam App ID
            const manifestFile = `appmanifest_${appId}.acf`;

            for (const libPath of uniqueLibraryPaths) {
                const manifestPath = path.join(libPath, 'steamapps', manifestFile);
                win.webContents.send('update-status', this.i18next.t('status_checking_steam_library', { library: libPath }));

                if (fs.existsSync(manifestPath)) {
                    const manifestContent = fs.readFileSync(manifestPath, 'utf-8');
                    const installDirMatch = manifestContent.match(/"installdir"\s+"([^"]+)"/);
                    if (installDirMatch && installDirMatch[1]) {
                        const installDir = installDirMatch[1];
                        const gameExePath = path.join(libPath, 'steamapps', 'common', installDir, 'BlueArchive.exe');
                        if (fs.existsSync(gameExePath)) {
                            win.webContents.send('update-status', this.i18next.t('status_found_steam', { path: gameExePath }));
                            await new Promise(resolve => setTimeout(resolve, 1500));
                            return gameExePath;
                        }
                    }
                }
            }
        } catch (err) {
            console.error('在搜尋 Steam 遊戲過程中發生錯誤:', err);
        }

        return null;
    }

    /**
     * 獲取系統磁碟機列表
     */
    async getDriveLetters() {
        try {
            const drives = await drivelist.list();
            console.log('--- Drivelist Raw Output ---');
            console.log(drives);
            const drivePaths = drives
                .filter(drive => drive.mountpoints && drive.mountpoints.length > 0)
                .flatMap(drive => drive.mountpoints.map(mp => `${mp.path.replace(/\\/g, '/')}/`));

            console.log('--- Filtered Drive Paths for Scanning ---');
            console.log(drivePaths);
            return drivePaths;
        } catch (error) {
            console.error('Failed to get drive letters with drivelist:', error);
            return ['C:/', 'D:/', 'E:/'];
        }
    }

    /**
     * 查找遊戲執行檔
     */
    async findGameExecutable(win) {
        try {
            const steamGamePath = await this.findGameViaSteam(win);
            if (steamGamePath) {
                this.saveGamePaths(steamGamePath);
                win.webContents.send('update-status', this.i18next.t('status_found', { path: steamGamePath }));
                return { gamePath: this.store.get('gamePath'), gameBundlePath: this.store.get('gameBundlePath') };
            }
            win.webContents.send('update-status', this.i18next.t('status_steam_not_found_fallback'));
            await new Promise(resolve => setTimeout(resolve, 2000));
        } catch (err) {
            console.error('Steam 搜尋過程中發生錯誤:', err);
        }

        win.webContents.send('update-status', this.i18next.t('status_preparing_search'));
        const searchPaths = await this.getDriveLetters();
        win.webContents.send('update-status', this.i18next.t('status_drives_found', { drives: searchPaths.join(', ') }));

        await new Promise(resolve => setTimeout(resolve, 1000));

        for (const searchPath of searchPaths) {
            try {
                win.webContents.send('update-status', this.i18next.t('status_scanning_drive', { drive: searchPath }));

                const entries = await fg('**/BlueArchive.exe', {
                    cwd: searchPath,
                    deep: 7,
                    onlyFiles: true,
                    caseSensitiveMatch: false,
                    suppressErrors: true,
                    ignore: [
                        '**/$RECYCLE.BIN/**',
                        '**/System Volume Information/**',
                        '**/Windows/**',
                        '**/ProgramData/**',
                        '**/$WinREAgent/**',
                        '**/Recovery/**'
                    ],
                });

                if (entries.length > 0) {
                    const foundPath = path.join(searchPath, entries[0]);
                    this.saveGamePaths(foundPath);
                    win.webContents.send('update-status', this.i18next.t('status_found', { path: foundPath }));
                    return { gamePath: this.store.get('gamePath'), gameBundlePath: this.store.get('gameBundlePath') };
                }
            } catch (err) {
                console.error(`Error searching in ${searchPath}:`, err);
            }
        }

        win.webContents.send('update-status', this.i18next.t('status_not_found'));
        return null;
    }

    /**
     * 保存遊戲路徑到設定中
     */
    saveGamePaths(executablePath) {
        const gameDirectory = path.dirname(executablePath);
        const bundlePath = path.join(gameDirectory, 'BlueArchive_Data');
        this.store.set({ gamePath: executablePath, gameBundlePath: bundlePath });
        return { gamePath: executablePath, gameBundlePath: bundlePath };
    }

    /**
     * 遞迴搜尋檔案
     */
    async findFileRecursively(directory, fileNameToFind) {
        try {
            const items = await fs.promises.readdir(directory, { withFileTypes: true });
            for (const item of items) {
                const fullPath = path.join(directory, item.name);
                if (item.isDirectory()) {
                    const result = await this.findFileRecursively(fullPath, fileNameToFind);
                    if (result) {
                        return result;
                    }
                } else if (item.name.toLowerCase() === fileNameToFind.toLowerCase()) {
                    return fullPath;
                }
            }
        } catch (err) {
        }
        return null;
    }

    /**
     * 查找目標檔案
     */
    async findTargetFile(modFileName, win) {
        const gamePath = this.store.get('gamePath');
        if (!gamePath) {
            console.warn('Game path not set in store.');
            return null;
        }

        const searchBase = path.join(path.dirname(gamePath), 'BlueArchive_Data');
        if (!fs.existsSync(searchBase)) {
            console.warn(`Search base directory does not exist: ${searchBase}`);
            return null;
        }

        win.webContents.send('update-action-status', this.i18next.t('status_searching_for_file', { file: modFileName }));
        console.log(`Searching for file: ${modFileName} in base directory: ${searchBase}`);

        const foundPath = await this.findFileRecursively(searchBase, modFileName);

        if (foundPath) {
            console.log(`File found: ${foundPath}`);
            return foundPath;
        }

        console.warn(`File not found: ${modFileName} in directory: ${searchBase}`);
        return null;
    }
}

module.exports = GamePathManager;

File: Blue Archive Mod Manager-C#/Modules/GamePathManager.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Win32;
using Blue_Archive_Mod_Manager_C_.Utils;

namespace Blue_Archive_Mod_Manager_C_.Modules
{
    public class GamePathManager
    {
        private readonly SettingsManager _settingsManager;

        public GamePathManager(SettingsManager settingsManager)
        {
            _settingsManager = settingsManager;
        }

        public string GetSteamInstallPath()
        {
            try
            {
                // Try 64-bit registry first
                using (var key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\WOW6432Node\Valve\Steam"))
                {
                    if (key?.GetValue("InstallPath") is string path64)
                        return path64;
                }

                // Try 32-bit registry
                using (var key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Valve\Steam"))
                {
                    if (key?.GetValue("InstallPath") is string path32)
                        return path32;
                }
            }
            catch
            {
                // Registry access failed
            }

            return null!;
        }

        public async Task<string> FindGameViaSteamAsync(Action<string>? statusCallback = null)
        {
            statusCallback?.Invoke("Finding Steam installation...");
            
            var steamPath = GetSteamInstallPath();
            if (string.IsNullOrEmpty(steamPath))
            {
                statusCallback?.Invoke("Steam not found in registry.");
                return null;
            }

            statusCallback?.Invoke($"Steam found at: {steamPath}");
            await Task.Delay(1000);

            var libraryFoldersPath = Path.Combine(steamPath, "steamapps", "libraryfolders.vdf");
            if (!File.Exists(libraryFoldersPath))
            {
                statusCallback?.Invoke("libraryfolders.vdf not found.");
                return null;
            }

            var libraryPaths = new List<string> { steamPath };
            
            try
            {
                var content = File.ReadAllText(libraryFoldersPath);
                var pathMatches = System.Text.RegularExpressions.Regex.Matches(content, @"""path""\s+""([^""]+)""");
                
                foreach (System.Text.RegularExpressions.Match match in pathMatches)
                {
                    var libPath = match.Groups[1].Value.Replace("\\\\", "\\");
                    if (Directory.Exists(libPath))
                    {
                        libraryPaths.Add(libPath);
                    }
                }
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error reading libraryfolders.vdf: {ex.Message}");
            }

            var appId = "3557620"; // Blue Archive Steam App ID
            var manifestFile = $"appmanifest_{appId}.acf";

            foreach (var libPath in libraryPaths)
            {
                statusCallback?.Invoke($"Checking Steam library: {libPath}");
                
                var manifestPath = Path.Combine(libPath, "steamapps", manifestFile);
                if (!File.Exists(manifestPath)) continue;

                try
                {
                    var manifestContent = File.ReadAllText(manifestPath);
                    var match = System.Text.RegularExpressions.Regex.Match(manifestContent, @"""installdir""\s+""([^""]+)""");
                    
                    if (match.Success)
                    {
                        var installDir = match.Groups[1].Value;
                        var gameExePath = Path.Combine(libPath, "steamapps", "common", installDir, "BlueArchive.exe");
                        
                        if (File.Exists(gameExePath))
                        {
                            statusCallback?.Invoke($"Found Blue Archive at: {gameExePath}");
                            await Task.Delay(1500);
                            return gameExePath;
                        }
                    }
                }
                catch (Exception ex)
                {
                    Console.Error.WriteLine($"Error checking manifest: {ex.Message}");
                }
            }

            statusCallback?.Invoke("Blue Archive not found in Steam libraries.");
            return null;
        }

        public List<string> GetDriveLetters()
        {
            try
            {
                return DriveInfo.GetDrives()
                    .Where(d => d.IsReady)
                    .Select(d => d.Name)
                    .ToList();
            }
            catch
            {
                return new List<string> { "C:\\", "D:\\", "E:\\" };
            }
        }

        public async Task<string> FindGameExecutableAsync(string region, Action<string>? statusCallback = null)
        {
            try
            {
                if (region != null && region.Equals("jp", StringComparison.OrdinalIgnoreCase))
                {
                    statusCallback?.Invoke("Starting JP Game Detection...");
                    var jpPath = await FindGameViaJpLauncherAsync(statusCallback);
                    if (!string.IsNullOrEmpty(jpPath))
                    {
                        SaveGamePaths(jpPath);
                        return jpPath;
                    }
                     // If JP specific detection fails, should we fall back to generic drive search?
                     // Yes, but we might find Global version. 
                     // Let's assume the user knows what they are doing if they selected JP.
                     statusCallback?.Invoke("JP Launcher method failed. Searching drives...");
                }
                else
                {
                    // Global (Steam) detection
                    var steamGamePath = await FindGameViaSteamAsync(statusCallback);
                    if (!string.IsNullOrEmpty(steamGamePath))
                    {
                        SaveGamePaths(steamGamePath);
                        return steamGamePath;
                    }
                    statusCallback?.Invoke("Steam search failed. Searching drives...");
                }
                
                await Task.Delay(2000);
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Search error: {ex.Message}");
            }

            statusCallback?.Invoke("Preparing drive search...");
            var drives = GetDriveLetters();
            statusCallback?.Invoke($"Found drives: {string.Join(", ", drives)}");
            
            await Task.Delay(1000);

            foreach (var drive in drives)
            {
                try
                {
                    statusCallback?.Invoke($"Scanning drive: {drive}");
                    var result = SearchDriveForGame(drive, 7, statusCallback);
                    
                    if (!string.IsNullOrEmpty(result))
                    {
                        SaveGamePaths(result);
                        statusCallback?.Invoke($"Found: {result}");
                        return result;
                    }
                }
                catch (Exception ex)
                {
                    Console.Error.WriteLine($"Error searching {drive}: {ex.Message}");
                }
            }

            statusCallback?.Invoke("Blue Archive not found.");
            return null!;
        }

        public async Task<string> FindGameViaJpLauncherAsync(Action<string>? statusCallback = null)
        {
            try
            {
                var appData = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
                // User provided: AppData\Roaming\BlueArchive_JP_Gamelauncher\Local Storage\leveldb\000003.log
                var logPath = Path.Combine(appData, "BlueArchive_JP_Gamelauncher", "Local Storage", "leveldb", "000003.log");
                
                statusCallback?.Invoke($"Checking JP Launcher log: {logPath}");

                if (File.Exists(logPath))
                {
                    // Try to read the file and look for paths
                    // The file is likely binary (LevelDB), but might contain plain text paths.
                    // We'll read it as string and regex search for absolute paths ending in BlueArchive.exe
                    // Or generically, paths containing "BlueArchive"
                    
                    // Note: LevelDB log files might be locked if launcher is running.
                    string content = "";
                    try 
                    {
                        using (var fs = new FileStream(logPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
                        using (var sr = new StreamReader(fs))
                        {
                            content = await sr.ReadToEndAsync();
                        }
                    }
                    catch (Exception readEx)
                    {
                         statusCallback?.Invoke($"Failed to read log: {readEx.Message}");
                         // Fallback: Check if the folder exists, maybe standard install location?
                         // Standard Yostar install: C:\Program Files\BlueArchiveJP ?
                         return null;
                    }

                    // Regex for path
                    // Pattern: Drive letter, colon, backslash, characters, BlueArchive.exe
                    // Note: Paths in logs might be escaped or use forward slashes.
                    // Updated to handle directory paths seen in logs like "C:\\YostarGames\\BlueArchive_JP"
                    var patterns = new[] 
                    {
                        // Match specific exe path if present
                        @"([a-zA-Z]:[\\/](?:[^<>:""/\\|?*]+[\\/])+BlueArchive\.exe)",
                        // Match quoted directory paths with double backslashes (common in JSON/C strings)
                        @"([a-zA-Z]:\\\\(?:[^""\x00-\x1F]+))"
                    };

                    foreach (var pat in patterns)
                    {
                        var matches = System.Text.RegularExpressions.Regex.Matches(content, pat, System.Text.RegularExpressions.RegexOptions.IgnoreCase);
                        foreach (System.Text.RegularExpressions.Match match in matches)
                        {
                             var rawPath = match.Groups[1].Value;
                             var path = rawPath.Replace("\\\\", "\\").Replace("/", "\\"); // Normalize
                             
                             // If path matches a directory, append exe name
                             if (!path.EndsWith("BlueArchive.exe", StringComparison.OrdinalIgnoreCase))
                             {
                                 path = Path.Combine(path, "BlueArchive.exe");
                             }

                             if (File.Exists(path))
                             {
                                 statusCallback?.Invoke($"Found JP Game at: {path}");
                                 return path;
                             }
                        }
                    }
                    
                    statusCallback?.Invoke("No valid game path found in log file.");
                }
                else 
                {
                    statusCallback?.Invoke("JP Launcher log file not found.");
                }
            }
            catch (Exception ex)
            {
                 Console.Error.WriteLine($"JP Launcher search error: {ex.Message}");
            }
            
            return null;
        }

        private string SearchDriveForGame(string drivePath, int maxDepth, Action<string>? statusCallback = null)
        {
            try
            {
                var searchTasks = new Stack<(string path, int depth)>();
                searchTasks.Push((drivePath, 0));

                var excludedDirs = new[] 
                { 
                    "$RECYCLE.BIN", "System Volume Information", "Windows", 
                    "ProgramData", "$WinREAgent", "Recovery" 
                };

                while (searchTasks.Count > 0)
                {
                    var (currentPath, depth) = searchTasks.Pop();

                    if (depth > maxDepth) continue;

                    try
                    {
                        var files = Directory.GetFiles(currentPath, "BlueArchive.exe", System.IO.SearchOption.TopDirectoryOnly);
                        if (files.Length > 0)
                        {
                            return files[0];
                        }

                        foreach (var dir in Directory.GetDirectories(currentPath))
                        {
                            var dirName = Path.GetFileName(dir);
                            if (!excludedDirs.Contains(dirName, StringComparer.OrdinalIgnoreCase))
                            {
                                searchTasks.Push((dir, depth + 1));
                            }
                        }
                    }
                    catch
                    {
                // Skip directories we can't access
                    }
                }
            }
            catch
            {
                // Search failed
            }

            return null!;
        }

        public (string gamePath, string gameBundlePath) SaveGamePaths(string executablePath)
        {
            var gameDirectory = Path.GetDirectoryName(executablePath);
            var bundlePath = Path.Combine(gameDirectory, "BlueArchive_Data");
            
            _settingsManager.Set(new Dictionary<string, object>
            {
                { "gamePath", executablePath },
                { "gameBundlePath", bundlePath }
            });

            return (executablePath, bundlePath);
        }

        public (string gamePath, string gameBundlePath) GetGamePaths()
        {
            var gamePath = _settingsManager.Get<string>("gamePath");
            var bundlePath = _settingsManager.Get<string>("gameBundlePath");
            return (gamePath, bundlePath);
        }

        public string? FindTargetFile(string fileName, int maxDepth = 6)
        {
            var (_, bundlePath) = GetGamePaths();
            if (string.IsNullOrEmpty(bundlePath) || !Directory.Exists(bundlePath))
            {
                return null;
            }

            var direct = Path.Combine(bundlePath, fileName);
            if (File.Exists(direct)) return direct;

            var stack = new Stack<(string path, int depth)>();
            stack.Push((bundlePath, 0));
            var skipDirs = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
            {
                "Cache", "Logs", "Temp", "TempCache"
            };

            while (stack.Count > 0)
            {
                var (current, depth) = stack.Pop();
                if (depth > maxDepth) continue;

                try
                {
                    var files = Directory.GetFiles(current, fileName, SearchOption.TopDirectoryOnly);
                    if (files.Length > 0) return files[0];

                    foreach (var dir in Directory.GetDirectories(current))
                    {
                        var name = Path.GetFileName(dir);
                        if (skipDirs.Contains(name)) continue;
                        stack.Push((dir, depth + 1));
                    }
                }
                catch
                {
                    // ignore inaccessible dirs
                }
            }

            return null;
        }
    }
}
